Source/missiles.cpp

```diff
@@ -1527,42 +1527,31 @@ void __fastcall LoadMissileGFX(BYTE mi)
 
 void __cdecl InitMissileGFX()
 {
-    int mi;
+	int mi;
 
-    for (mi = 0; misfiledata[mi].mAnimFAmt; mi++) {
-        if (!(misfiledata[mi].mFlags & MFLAG_HIDDEN))
-            LoadMissileGFX(mi);
-    }
+	for (mi = 0; misfiledata[mi].mAnimFAmt; mi++) {
+		if (!(misfiledata[mi].mFlags & MFLAG_HIDDEN))
+			LoadMissileGFX(mi);
+	}
 }
 
 void __fastcall FreeMissileGFX(int mi)
 {
-	int v1;            // esi
-	unsigned char *v2; // ecx
-	signed int v3;     // ebx
-	void **v4;         // edi
-	void *v5;          // ecx
+	int i;
+	void *p;
 
-	v1 = mi;
 	if (misfiledata[mi].mFlags & MFLAG_ALLOW_SPECIAL) {
-		v2 = misfiledata[v1].mAnimData[0];
-		if (v2) {
-			mem_free_dbg(&v2[-4 * misfiledata[v1].mAnimFAmt]);
-			misfiledata[v1].mAnimData[0] = 0;
+		if (misfiledata[mi].mAnimData[0]) {
+			mem_free_dbg(&misfiledata[mi].mAnimData[0][-4 * misfiledata[mi].mAnimFAmt]);

```

Comment:
While bin exact, this looks rather strange doesn't it?

What does `-4 * misfiledata[mi].mAnimFAmt` signal? How is it supoesed to be interpreted?

Reply:
mAnimData is 16 byte arrays, so 4 is probably to switch to 32bit instead of 8bit. That and the fact that it's negative and acting on item 0 makes me think that it is actually trying to access on e the previous elements in the struct. mAnimFAmt has a range of 0-16.
Probably we can better understand what it's doing by lookin at the equvelant loading section:
```
		sprintf(pszName, "Missiles\\%s.CEL", mfd->mName);
		file = LoadFileInMem(pszName, 0);
		for (i = 0; i < mfd->mAnimFAmt; i++)
mfd->mAnimData[i] = &file[((int *)file)[i]];
```

Reply:
Basically this should be equivalent, though still not very clear:
```suggestion
			mem_free_dbg(&((int *)misfiledata[mi].mAnimData[0])[-misfiledata[mi].mAnimFAmt]);
```

Reply:
Hmm. I tried to have a deeper look but still don't quite understand what's happening. To much pointer arithmatic and magic :) Still, given it's bin exact we can merge it and simply try to make it prettier later. Perhaps one of us will have a stroke of insight.

Reply:
![image](https://user-images.githubusercontent.com/14297035/54631672-c4d1e180-4a7c-11e9-80ef-097f5da254d6.png)
a stroke of insight u say

Reply:
This one is a bit tricky. Missile files are stored as `.CL2`, therefore `mAnimData` contains a pointer to a table of pointers. It looks to me you'll have to do the same thing done in `engine.cpp`. You need to cast it to a `DWORD *` and then free the pointer of the pointer.

Reply:
Also, whoever cleaned up LoadMissileGFX broke it. The file extensions should be `.CL2` not `.CEL`

Reply:
me : P

Reply:
Found some relevant code in the new cleaned up Cl2DecodeFrm3 and applied it to this function. I think it is much more readable and ready to merge now. What do you think @mewmew ?

Reply:
Still, to me the negative integer used to index array elements sounds quite wrong:
`(... .mAnimData[0])[-misfiledata[mi].mAnimFAmt]`

(Not that my opinion is very well researched)

Reply:
> Still, to me the negative integer used to index array elements sounds quite wrong:
`(... .mAnimData[0])[-misfiledata[mi].mAnimFAmt]`

I would have to agree with @sskras, but also can't think of what code would be right in this case. Perhaps it just was a *really strange* way to write how to access the pointers of the allocated CL2 graphics? My intuition says that there probably was a field of a struct that was accessed, rather than this pointer magic. But I may be wrong.